cmd/device.go

```diff
@@ -136,27 +141,31 @@ func runDevicePinger(wg *sync.WaitGroup, ping client.AgentPing, apiOrigin string
 			handleDeviceUpdate(ping, config)
 		}
 
-		if config.Enabled && config.Host != "" {
-			// ping server instead of sleeping
-			pinger, err := goping.NewPinger(config.Host)
-			if err == nil {
-				log.V(2).Info("Pinging server", "host", config.Host)
-				pinger.Count = AgentPingInterval * 10
-				pinger.Interval = time.Millisecond * 100
-				pinger.Timeout = time.Second * AgentPingInterval
-				pinger.Run()
-				log.V(1).Info("Done pinging server", "host", config.Host, "stats", *pinger.Statistics())
-				getPingStats(&ping, pinger.Statistics())
-			} else {
-				log.Error(err, "Failed to create pinger")
-				// sleep in between pings
-				time.Sleep(time.Second * AgentPingInterval)
-				getPingStats(&ping, nil)
-			}
+		if config.Enabled == false || config.Host == "" {
+			// There is no host configured, so move on
+			time.Sleep(AgentPingInterval)
+			getPingStats(&ping)
 		} else {
-			// sleep in between pings
-			time.Sleep(time.Second * AgentPingInterval)
-			getPingStats(&ping, nil)
+			// establish a socket if there is no active connection
+			duration, errIfce := timeIt(func() interface{} {
+				u := fmt.Sprintf("%s:%d", config.Host, config.Port)
+				conn, err := net.DialTimeout("tcp", u, time.Duration(time.Second))
+				if err != nil {
+					return err
+				}
+				conn.Close()
+				return nil
+			})
+			if errIfce != nil {
+				log.Error(err, "Tcp connection interface error")
+				PingRecorder.Reset()
+				time.Sleep(AgentPingInterval)
+				continue
+			}
+			PingRecorder.addPingRecord(duration)
+			getPingStats(&ping)
+
+			time.Sleep(AgentPingInterval)
```

Comment:
Should you sleep before `getPingStats()`? It's probably minor, but the previous implementation would also sleep first. This might only impact the first start which might trigger 2x the pings in one interval?